src/platform_linux/shmem.c

```diff
@@ -742,42 +965,248 @@ platform_shm_track_alloc(shmem_info *shm, void *addr, size_t size)
+ * -----------------------------------------------------------------------------
+ */
+/*
+ * Simple lookup routine to return the free-fragment header off of which
+ * free-fragments of a specific 'size' will be hung off of.
+ * No mutex is required here, as we are simply mapping size to a field's addr.
+ */
+static free_frag_hdr **
+platform_shm_free_frag_hdr(const shmem_info *shm, size_t size)
+{
+   free_frag_hdr **next_frag;
+   if (size <= 64) {
+      next_frag = (free_frag_hdr **)&shm->shm_free_le64;
+   } else if (size <= 128) {
+      next_frag = (free_frag_hdr **)&shm->shm_free_le128;
+   } else if (size <= 256) {
+      next_frag = (free_frag_hdr **)&shm->shm_free_le256;
+   } else if (size <= 512) {
+      next_frag = (free_frag_hdr **)&shm->shm_free_le512;
+   } else {
+      // Currently unsupported fragment size for recycling
+      next_frag = NULL;
+   }
+   return next_frag;
+}
+
+/*
+ * When a memory fragment is being free'd, check if this fragment is already
+ * in some free-list. If found, it means we are [incorrectly] doing a
+ * double-free, which indicates a code error. User has possibly messed-up their
+ * handling of memfrag handles to this memory fragment.
+ *
+ * NOTE: This, being a convenience routine, provides for next_frag, which is
+ * the start of the free-list for given 'size'. If caller has established it,
+ * pass that here. Otherwise, we will establish it in this routine.
+ *
+ * Shared memory mutex is expected to be held for this function.
+ */
+static void *
+platform_shm_find_frag_in_free_list(const shmem_info *shm,
+                                    free_frag_hdr   **next_frag,
+                                    const void       *ptr,
+                                    const size_t      size)
+{
+   if (!next_frag) {
+      free_frag_hdr **next_frag = platform_shm_free_frag_hdr(shm, size);
+      // We are searching for a fragment whose size is not tracked.
+      if (next_frag == NULL) { // Nothing found.
+         return NULL;
+      }
+   }
+
+   // Walk the free-list to see if our being-free'd ptr lives there already
+   while (*next_frag && ((*next_frag) != ptr)) {
+      next_frag = &(*next_frag)->free_frag_next;
+   }
+   // Returns the 'ptr' if found; null otherwise.
+   return (*next_frag);
+}
+
+/*
+ * Diagnostic routine: Iterate through all free-lists that we currently
+ * manage and try to find if fragment at address 'ptr' is found in any such
+ * list. That means, fragment was previously freed.
+ *
+ * Returns: The size of the free-fragment list in which this 'ptr' was found.
+ *  0, otherwise; (i.e. 'ptr' is not an already-freed-fragment.)
+ *  Optionally, the size marked in this freed-fragment is returned via
+ *  'freed_size'. If a client incorrectly specified the memfrag's size at
+ *  the time of free(), that will be reported here, and can be detected.
+ */
+static size_t
+platform_shm_find_frag_in_freed_lists(const shmem_info *shm,
```

Comment:
Used for debug-asserts, and also a wrapper to this is provided, so test code can call it and do this deep cross-check.